### ISSUE TITLE
Possible improvement: cloud file caching

### ISSUE THREAD

**Issue by radeusgd** [enso-org/enso#11439]
Much like the caching for `Data.fetch` implemented in #11342, we could have a similar solution for when we are reading cloud files.

We cannot re-use the `HTTP` layer cache directly, because the pre-signed URL used to download the Cloud file is likely to change (because they expire). So the URLs would be different between calls.

But we can actually do better - when downloading the file, we are fetching the `get_details` endpoint to get the current pre-signed URL for download. That same endpoint could be made to report the file _version_ that is stored in the cloud. 

Thus we could implement our own cache that relies on the `asset_id` and cloud file version. Such cache would only require re-downloading the file if a new version was uploaded, and it will be more 'correct' than TTL-based one, because as soon as the file is updated, the version changes so any future downloads will correctly invalidate the cache entry.

**Comment by GregoryTravis** [comment]
I think this could be implented using LRUCache -- if you agree, feel free to assign this to me.

**Comment by GregoryTravis** [comment]
And it might be worthwhile to implement ETag checking for Data.fetch.

**Comment by radeusgd** [comment]
> I think this could be implented using LRUCache -- if you agree, feel free to assign this to me.

Yes, I think we should be able to re-use some of the LRUCache logic, although it may then require to abstract-away the TTL handling. Since the Enso cloud cache should not rely on TTL, but instead of the asset version reported by cloud (in discussion with cloud team if we can get it on the `get_file_details` endpoint).

**Comment by radeusgd** [comment]
The ticket adding the version number in the cloud endpoint: https://github.com/enso-org/cloud-v2/issues/1568

**Comment by farmaazon** [comment]
The above was meant for #11493
